### ISSUE TITLE
Move docket view count out of docket model

### ISSUE THREAD

**Issue by mlissner** [freelawproject/courtlistener#3346]
Right now we tally a count to the docket table every time somebody views a docket. I'm sorry to say this is a bad idea. Looking up, locking, and then incrementing a row on a table of this size every time somebody loads the page is difficult. In fact, I've noticed that this query is our worst query according to AWS:

![image](https://github.com/freelawproject/courtlistener/assets/236970/7a1ecbb8-18c7-4a73-968a-b3fcb67a9282)

(You can't really tell, but it's the first one you see there.)

I'm not sure the best fix for this: 
 - We could put this in redis. 
 - We could use our stats app for it. 
 - We could develop something new. 
 - We could use some kind of external API like Plausible or whatever (this seems really slow). 
 - We could use a special stats database for this (Clickhouse?)

We should do something though, I'm afraid....

**Comment by mlissner** [comment]
By the way, the #917 will also fix this problem when we do it (at the cost of losing the accuracy of this stat).

**Comment by mlissner** [comment]
We're using `view_count` to generate sitemaps, so it might not be so easy to remove unless we think of a better thing to use for that. One way to do this better is outlined here: 

https://medium.com/@timanovsky/ultra-fast-asynchronous-counters-in-postgres-44c5477303c3

Basically, rather than incrementing the counter in realtime, you do an insert in some other table. Then, every minute or so a cronjob adds up the rows in the table and increments whichever view_count fields are needed. Interesting approach, and leaves the value in the DB, which lets us continue querying it. Could also be done using redis and a cronjob instead of inserting rows.

When the CDN is in place, I guess we'll need a very fast API for this so that the client can increment the caller.

**Comment by mlissner** [comment]
Note that this also affects the `citation_count` field on `OpinionCluster`s.

**Comment by mlissner** [comment]
- [ ] We need to mark the view_count field as deprecated wherever it is in the DB and make an issue to remove it eventually.

**Comment by mlissner** [comment]
Hm, but we don't need _unique_ counts on each page — or am I lacking imagination of how a HLL could be useful to us?

> Moving the view count to Redis would involve writing additional code to manage the data, specifically for cleaning up Redis and then saving those counts to the database.

Do we need to save these to the DB? What for? 

> We also need to decide if an exact count is important.

It's not, I don't think.

**Comment by ERosendo** [comment]
> we don't need unique counts on each page

You're correct

> or am I lacking imagination of how a HLL could be useful to us?

I thought we could use HLL by combining the request IP with a timestamp — something like `ip:timestamp` — so that each page view appears unique. That way, even repeat visits from the same IP would register as separate entries, which would inflate the HLL count and make it behave more like a total view counter. It’s definitely a bit of a hack — the counts would be approximate — but it could work if we wanted a very lightweight, memory-efficient estimate of total views without needing exact accuracy.

> Do we need to save these to the DB? What for?

I figured we wouldn’t want to keep a separate HLL structure in Redis for every docket indefinitely — especially since Redis is in-memory and and we're already storing view counts in the database. (We'd also tweak the DB schema to move the count to a smaller table, preventing large dead tuples when updating rows). So the idea was to periodically extract the estimated cardinality from Redis and store that value in the DB. That way, we can clean up Redis regularly and avoid accumulating too many keys, while still retaining view data.

**Comment by mlissner** [comment]
One other thing I'm remembering is that we do query the DB for "popular" dockets so they can go in our sitemaps (otherwise, there are just too many things in the sitemaps), so putting the value in the DB is useful. 

The advantage of redis is that it is a bit more performant, but I think a small two-column table for this in the DB should be fine. I imagine postgres would put the table (or parts of it) into memory (or virtual memory) anyway.

One other performance thing we could do is to only tally 1/10 requests, like we do with Plausible, then always remember to multiply × 10 when we look at the numbers. This would eliminate the performance concerns of this, but is kind of annoying and probably not needed — so I'd say let's put this in the DB, hope for the best, and if it's an issue later, rethink it again. 


-----

One last thought is that if we're making this generic, an FK to the docket table might not be quite what we want. Like, in the future, we'll probably want to count views of something else or something other than views. 

Would it work to do a two-column table with no FK, where the first column is something sort of like you'd put in redis: 

```
d.view:$id
```

And the second is the count, so for the first view of docket 1234, you'd do:

```
d.view:1234,1
```

And the first view count for an opinion could be:

```
o.view:456,1
```

I worry that FK's have secret magic we'd be going without, but this feels like it might be a nice way to do this generically?

**Comment by ERosendo** [comment]
> Would it work to do a two-column table with no FK, where the first column is something sort of like you'd put in redis: 
> 
> ```
> d.view:$id
> ```
> 
> And the second is the count, so for the first view of docket 1234, you'd do:
> 
> ```
> d.view:1234,1
> ```
> 
> And the first view count for an opinion could be:
> 
> ```
> o.view:456,1
> ```

Sounds like a polymorphic relationship, right? In Django, the go-to for this kind of scenario is often a [Generic relation](https://docs.djangoproject.com/en/5.2/ref/contrib/contenttypes/#generic-relations )

**Comment by mlissner** [comment]
Yeah, I wondered if that might be the way to do it, but I hesitated because I can't remember the tradeoffs of using the ContentTypes system. I have a vague memory that it's hard to work with, performs badly, migrates badly, or something like that? I really can't remember the tradeoff though — maybe there isn't one? Sorry.

If we do decide to use that system, we won't be able to count other stuff, but maybe that's fine.
